### PR TITLE
Implement refreshing the Cloud token in Enso libraries

### LINKED ISSUE #9300: Refresh Cloud access token when needed to avoid 401 errors in our cloud integration
With #9287 we are getting access to a 'refresh token' that can be used to refresh the access token whenever it expires.

- [x] Extend the logic reading the token to read in, and store, the additional metadata.
- [x] Implement a method `Utils.refresh_access_token` that negotiates a refreshed token and saves it in `AuthenticationProvider`'s cache.
	- [x] Note that this may _still_ fail if the refresh token itself is also expired (e.g. a month of inactivity). We should handle that case and raise an informative error to the user (e.g. "Your credentials have expired. Please log-in to the Cloud again and restart the workflow.").
- [x] Whenever `authorization_header` is accessed, check if the access token might have expired.
	- [x] If it has expired, automatically renegotiate a new one using `refresh_access_token` before returning.
- We can still have an edge case where the token has not expired yet when `authorization_header` is fetched just before the token is expired, and once the actual request is made, the token already expires.
    - To make this less likely, let's be more eager with token refresh - e.g. if there's less than 60 seconds left on the token, we should refresh it already.
    - [x] We should consider if we should have a custom decorator for our requests that catches `401` error, refreshes the token and re-runs the whole request. 
	    - This is done by making `Enso_Cloud.Internal.Utils.http_request` retry the request. All of our Enso Cloud requests should be going through this method.
- [x] Given that this logic is relatively brittle and some edge cases may occur very rarely in day-to-day use, to make sure all of this works, I think it may be best to emulate a very basic auth flow in our HTTP test server to be able to run these various edge case scenarios:
   - Refreshing the token.
   - Refresh token itself being expired.
   - Request failing with 401 because the access token has just expired between `authorization_header` was fetched vs the request was issued + recovering from that.

### PR DESCRIPTION
**Pull request by radeusgd**
### Pull Request Description

- Closes #9300 
- Now the Enso libraries are themselves capable of refreshing the access token, thus there is no more problems if the token expires during a long running workflow.
- Adds `get_optional_field` sibling to `get_required_field` for more unified parsing of JSON responses from the Cloud.
	- Adds `expected_type` that checks the type of extracted fields. This way, if the response is malformed we get a nice Enso Cloud error telling us what is wrong with the payload instead of a `Type_Error` later down the line.
- Fixes `Test.expect_panic_with` to actually catch only panics. Before it used to also handle dataflow errors - but these have `.should_fail_with` instead. We should distinguish these scenarios.

### Checklist

Please ensure that the following checklist has been satisfied before submitting the PR:

- [x] The documentation has been updated, if necessary.
- [ ] Screenshots/screencasts have been attached, if there are any visual changes. For interactive or animated visual changes, a screencast is preferred.
- [x] All code follows the
      [Scala](https://github.com/enso-org/enso/blob/develop/docs/style-guide/scala.md),
      [Java](https://github.com/enso-org/enso/blob/develop/docs/style-guide/java.md),
      and
      [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md)
      style guides. In case you are using a language not listed above, follow the [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md) style guide.
- All code has been tested:
  - [x] Unit tests have been written where possible.
  - [ ] If GUI codebase was changed, the GUI was tested when built using `./run ide build`.

### CODE REVIEW

File: app/gui/view/graph-editor/src/builtin/visualization/java_script/table.js
Comment (radeusgd) [inline app/gui/view/graph-editor/src/builtin/visualization/java_script/table.js:1]: I removed this file because it was empty and it was causing problems on my machine.

That's why this PR has triggered the GUI codeowners.

File: std-bits/base/src/main/java/org/enso/base/enso_cloud/CloudAPI.java [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+    if (directoryId == null) {
+      // No current working directory is set
+      return null;
+    }
+
+    // TODO we should be able to fetch the name and organizationId from the cloud:
+    String directoryName = "???";
+    String organizationId = "";
+    cachedWorkingDirectory = new CloudWorkingDirectory(directoryName, directoryId, organizationId);
+    return cachedWorkingDirectory;
```
Comment (radeusgd) [inline std-bits/base/src/main/java/org/enso/base/enso_cloud/CloudAPI.java]: This is not new code, just moved during a refactor.
I guess it should be handled around #9289

File: distribution/lib/Standard/Base/0.0.0-dev/src/Enso_Cloud/Errors.enso [hunk 2]
```diff
@@ -5,10 +5,23 @@ import project.Network.HTTP.HTTP_Status_Code.HTTP_Status_Code
 ## PRIVATE
    Error if the user is not logged into Enso Cloud.
 type Not_Logged_In
+    ## PRIVATE
+    Error
```
Comment (GregoryTravis) [inline distribution/lib/Standard/Base/0.0.0-dev/src/Enso_Cloud/Errors.enso:9]: Why is it preferred to have a constructor?

File: distribution/lib/Standard/Base/0.0.0-dev/src/Enso_Cloud/Internal/Authentication.enso [hunk 3]
```diff
@@ -0,0 +1,166 @@
...
+
+    ## PRIVATE
+    read_from_credentials -> Authentication_Data =
+        content = credentials_file.read_text . catch File_Error error-> case error of
+            File_Error.Not_Found _ -> Panic.throw Not_Logged_In.Error
+            _ -> Panic.throw (Illegal_State.Error "Failed to read credentials file: "+error.to_display_text cause=error)
+        as_json = content.parse_json.catch Invalid_JSON error->
+            Panic.throw (Illegal_State.Error "Invalid credentials file format: "+error.to_display_text cause=error)
+        get_field field_name =
+            as_json.get field_name if_missing=(Panic.throw (Illegal_State.Error "Invalid credentials file format: missing field `"+field_name+"`."))
```
Comment (GregoryTravis) [inline distribution/lib/Standard/Base/0.0.0-dev/src/Enso_Cloud/Internal/Authentication.enso]: This could be shared with the similar code in `Refresh_Token_Data`, passing in a prefix for error messages.

**Review by GregoryTravis — APPROVED** [review]

File: distribution/lib/Standard/Base/0.0.0-dev/src/Enso_Cloud/Internal/Utils.enso [hunk 4]
```diff
@@ -86,42 +60,53 @@ datalinks_api = cloud_root_uri + "connectors"
...
-            handler = error_code.if_not_nothing <| error_handlers.get error_code
-            case handler of
-                Nothing -> Error.throw (Enso_Cloud_Error.Unexpected_Service_Error response.code payload)
-                _ : Function -> handler json_payload
+    if response.is_error && (retries > 0) then http_request method url body additional_headers error_handlers (retries - 1) else
+        if response.code.is_success then response else
+            is_unauthorized = (response.code == HTTP_Status_Code.unauthorized) || (response.code == HTTP_Status_Code.forbidden)
+            case is_unauthorized of
+                True -> if retries < 0 then Error.throw Enso_Cloud_Error.Unauthorized else
+                    Authentication.refresh_access_token
```
Comment (jdunkerley) [inline distribution/lib/Standard/Base/0.0.0-dev/src/Enso_Cloud/Internal/Utils.enso:100]: Feels wrong to run the refresh 3 times? Small detail but seems like should fail if unauthorized and has done the refresh token.

File: app/gui/view/graph-editor/src/builtin/visualization/java_script/table.js
Comment (jdunkerley) [inline app/gui/view/graph-editor/src/builtin/visualization/java_script/table.js:1]: why has this been touched here?

Comment (farmaazon) [inline app/gui/view/graph-editor/src/builtin/visualization/java_script/table.js:1]: Some leftover from the old GUI? It caught me as reviewer to this PR.

**Review by jdunkerley — APPROVED** [review]

File: distribution/lib/Standard/Base/0.0.0-dev/src/Enso_Cloud/Errors.enso (hunk 2 above)
Comment (radeusgd) [reply distribution/lib/Standard/Base/0.0.0-dev/src/Enso_Cloud/Errors.enso:9]: Without the constructor the `to_display_text` method did not work correctly.

Singleton types are not specified well enough IMO in current Enso implementation and method dispatch on them is a bit wonky - I guess I could also remove the `self` argument from `to_display_text` to make it a 'static' method, but I was less sure that this solution will work right in all contexts, so it felt safer to just use the same pattern we use elsewhere - a regular atom type instead of singleton.

File: app/gui/view/graph-editor/src/builtin/visualization/java_script/table.js
Comment (radeusgd) [reply app/gui/view/graph-editor/src/builtin/visualization/java_script/table.js:1]: That seems to be the case. [As mentioned](https://github.com/enso-org/enso/pull/9390#discussion_r1522073920) it was just an empty file that was causing compilation problems for me, so I removed it.

@farmaazon can you approve this change?

Comment (radeusgd) [reply app/gui/view/graph-editor/src/builtin/visualization/java_script/table.js:1]: https://github.com/enso-org/enso/pull/9390#discussion_r1522073920

File: distribution/lib/Standard/Base/0.0.0-dev/src/Enso_Cloud/Internal/Utils.enso (hunk 4 above)
Comment (radeusgd) [reply distribution/lib/Standard/Base/0.0.0-dev/src/Enso_Cloud/Internal/Utils.enso:100]: What other approach would you suggest?

I don't see any simpler ones, but I'm open to suggestions.

Currently the retry logic retries up to 3 times regardless of the kind of error.

What if the 1st retry fails due to network issue, second needs token refresh but also fails due to 500 server error and only the 3rd one succeeds? I don't see a difference between retrying for different reasons.

Note that if the 'refresh token' operation itself fails - that has no retry so it will fail regardless of the retry count. Shall I add the retry there too maybe?

File: distribution/lib/Standard/Base/0.0.0-dev/src/Enso_Cloud/Internal/Authentication.enso (hunk 3 above)
Comment (radeusgd) [reply distribution/lib/Standard/Base/0.0.0-dev/src/Enso_Cloud/Internal/Authentication.enso]: IMO it's such a small piece of code that abstracting it and adding the prefix makes the code more complicated, I did not share it on purpose.

File: app/gui/view/graph-editor/src/builtin/visualization/java_script/table.js
Comment (farmaazon) [reply app/gui/view/graph-editor/src/builtin/visualization/java_script/table.js:1]: Ah, so it is being removed from repo, not added (GH didn't display usual message "hiding because file is deleted" message, what confused me).

**Review by farmaazon — APPROVED** [review]

File: distribution/lib/Standard/Base/0.0.0-dev/src/Enso_Cloud/Internal/Authentication.enso (hunk 3 above)
Comment (radeusgd) [reply distribution/lib/Standard/Base/0.0.0-dev/src/Enso_Cloud/Internal/Authentication.enso]: After second thought I will try to improve the error handling a bit.

Comment (radeusgd) [reply distribution/lib/Standard/Base/0.0.0-dev/src/Enso_Cloud/Internal/Authentication.enso]: Refactored in https://github.com/enso-org/enso/pull/9390/commits/d87232886ab733cc267d0f0ed55bc1cd980f2eb2

File: engine/runtime/src/main/java/org/enso/interpreter/runtime/builtin/Error.java [hunk 5]
```diff
@@ -184,7 +184,7 @@ public Atom makeNoConversionCurrying(
    */
   @CompilerDirectives.TruffleBoundary
   public Atom makeTypeError(Object expected, Object actual, String name) {
-    return typeError.newInstance(expected, actual, Text.create("`" + name + "`"));
+    return typeError.newInstance(expected, actual, Text.create("Expected `" + name + "` to be {exp}, but got {got}."));
```
Comment (JaroslavTulach) [inline engine/runtime/src/main/java/org/enso/interpreter/runtime/builtin/Error.java]: Ajajaj, omission.

**Review by JaroslavTulach — APPROVED** [review]

File: test/Base_Tests/src/Data/Array_Proxy_Spec.enso [hunk 6]
```diff
@@ -43,7 +43,7 @@ add_specs suite_builder =
 
         group_builder.specify "should check validity at construction" <|
             Test.expect_panic_with (Array_Proxy.new 0 0) Type_Error
-            Test.expect_panic_with (Array_Proxy.new -1 (x->x)) Illegal_Argument
+            (Array_Proxy.new -1 (x->x)).should_fail_with Illegal_Argument
```
Comment (radeusgd) [inline test/Base_Tests/src/Data/Array_Proxy_Spec.enso:46]: I have now fixed the `Test.expect_panic_with` to only intercept **actual Panics**. Before it used to happily catch dataflow errors as well and not distinguish between them.

Our tests should distinguish between the two cases.

After my changes this test used to fail with:
image
